### PR TITLE
fix maxpool2d for XLA dynamo tracing

### PR DESCRIPTION
**Pull request by bdhirsh**
waiting for CI before getting review

### CODE REVIEW

**Comment by bdhirsh** [comment]
@JackCaoG do you know why the option to re-run is grey'd out on the CI failure? Not sure if it's a permissions thing or something else. It looks like CI didn't pick up my torch pin or something, trying to kick it off again:
image

**Comment by JackCaoG** [comment]
hmm I think torch pin has taken effect
```
+ TORCH_PIN=/tmp/pytorch/xla/scripts/../torch_patches/.torch_pin
+ '[' -f /tmp/pytorch/xla/scripts/../torch_patches/.torch_pin ']'
++ cat /tmp/pytorch/xla/scripts/../torch_patches/.torch_pin
+ CID='#90226'
+ [[ #90226 = \#* ]]
+ PRNUM=90226
+ set +x
Fetching PyTorch PR #90226
/tmp/pytorch /tmp/pytorch
From https://github.com/pytorch/pytorch
 * [new ref]               refs/pull/90226/head -> 90226
Switched to branch '90226'
M	third_party/ideep
M	third_party/kineto
Submodule path 'third_party/ideep': checked out 'ececd0a4f53c39f2d91caaddee0de1cd214f5b99'
Submodule path 'third_party/kineto': checked out '0703c78999061b8329dfab7ec5046fc5764a5573'
```
in the full log

**Comment by bdhirsh** [comment]
@JackCaoG I'm trying to rebuild XLA locally on my new devserver (I'm anticipating issues) - but I just stared at the code for a while and I think I know what the problem is. Just pushed it, so I'll see what the latest round of CI yields.

@shunting314 I'll give you a shout when this PR looks ready to test - when it is, can you try re-running your dynamo-XLA integration with max_pool2d (both fw and bw) and confirm if there are issues?

**Comment by shunting314** [comment]
@bdhirsh  sure, I'd be glad to do the tests

**Comment by bdhirsh** [comment]
Hey @shunting314, it looks like the max_pool2d unit tests are passing. I do see a failure in the XLA-dynamo tests, but it doesn't seem related to this change (?)

```
ERROR: test_simple_model (__main__.DynamoBasicTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/pytorch/xla/test/dynamo/test_dynamo.py", line 41, in test_simple_model
    res_xla_dynamo_2 = self.fn_simple_dynamo(xla_x, xla_y)
  File "/opt/conda/lib/python3.7/site-packages/torch/_dynamo/eval_frame.py", line 209, in _fn
    return fn(*args, **kwargs)
  File "/tmp/pytorch/xla/test/dynamo/test_dynamo.py", line 21, in fn_simple_dynamo
    @dynamo.optimize('torchxla_trace_once')
  File "/opt/conda/lib/python3.7/site-packages/torch/_dynamo/eval_frame.py", line 209, in _fn
    return fn(*args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/_dynamo/optimizations/backends.py", line 755, in fwd
    model = subgraph.model
NameError: free variable 'subgraph' referenced before assignment in enclosing scope
```

**Comment by JackCaoG** [comment]
rebase should fix the issue

**Comment by JackCaoG** [comment]
@bdhirsh you might also need to rebase your pytorch pr.

**Comment by bdhirsh** [comment]
yep - done

**Comment by shunting314** [comment]
@bdhirsh  is this the only PR i need patch? ( I previously saw you have 2 related PRs?)

**Comment by shunting314** [comment]
Do I need patch ' https://github.com/pytorch/pytorch/pull/90226 ' as well ?

**Comment by bdhirsh** [comment]
yes sorry - you'll need both :)

**Comment by shunting314** [comment]
@bdhirsh  I seed the following errors when building torchxla

```
# BUILD_CPP_TESTS=0 python setup.py develop
No CUDA runtime is found, using CUDA_HOME='/usr/local/cuda'
Building torch_xla version: 1.14
XLA Commit ID: c6acdf5417e736429e353e7abf2a26cde1d96122
PyTorch Commit ID: 9f04175b952abbe841e52e65fc95c0bd61a2f2cd
Traceback (most recent call last):
  File "/pytorch/xla/scripts/gen_lazy_tensor.py", line 120, in <module>
    get_device_fn="torch_xla::bridge::GetXlaDevice")
  File "/pytorch/torchgen/gen_lazy_tensor.py", line 367, in run_gen_lazy_tensor
    source_yaml, grouped_native_functions, backend_indices
  File "/pytorch/torchgen/gen_backend_stubs.py", line 246, in parse_backend_yaml
    {forward_kernels[0].kernel} is listed under "supported", but {backward_kernels[0].kernel} is listed under "autograd".'
AssertionError: Currently, all variants of an op must either be registered to a backend key, or to a backend's autograd key. They cannot be mix and matched. If this is something you need, feel free to create an issue! max_pool2d is listed under "supported", but max_pool2d is listed under "autograd".
Failed to generate lazy files: ['python', '/pytorch/xla/scripts/gen_lazy_tensor.py']
```

**Comment by shunting314** [comment]
oh, nvm, let my patch the pytorch side PR as well

**Comment by shunting314** [comment]
@bdhirsh I still see the issue after patching this PR and the corresponding PR on pytorch side. I've created a standalone tests without the need of patching my PR. You can repro using this simple script in your environment: https://github.com/pytorch/torchdynamo/issues/1837#issuecomment-1312287846

File: torch_xla/csrc/aten_autograd_ops.cpp [hunk 1]
```diff
@@ -99,11 +104,26 @@ torch::Tensor MaxPool2dAutogradFunction::forward(
...
+  // that XLA could reasonably run into,
+  // and mask them with the current tensor's keyset.
+  auto mask = c10::DispatchKeySet({
+    c10::DispatchKey::XLA,
+    c10::DispatchKey::Python,
+    c10::DispatchKey::Functionalize,
+    c10::DispatchKey::Conjugate,
+    c10::DispatchKey::Negative,
+    c10::DispatchKey::ZeroTensor,
+    c10::DispatchKey::ADInplaceOrView,
```
Comment (bdhirsh) [inline torch_xla/csrc/aten_autograd_ops.cpp:120]: Oh we should take out ADInplaceOrView from here

**Comment by bdhirsh** [comment]
Hey @JackCaoG - do you mind finishing up landing? @shunting314 confirmed that this fixes the E2E tests for max_pool2d

**Review by JackCaoG — APPROVED** [review]
Thanks!

**Pull request merged**